gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -198,30 +207,41 @@ private Object doResumableUpload() throws IOException, InterruptedException {
         stub.withDeadlineAfter(WRITE_STREAM_TIMEOUT.toMillis(), TimeUnit.MILLISECONDS)
             .insertObject(responseObserver);
         responseObserver.done.await();
-        writeOffset += responseObserver.bytesWritten();
-      } while (!responseObserver.isFinished());
 
-      if (responseObserver.hasError()) {
-        // TODO(b/150892988): Support resuming an upload after a transient error as follows:
-        //  1. Build a wrapper class around the ByteString (or subclass it), and keep track of the
-        //     start offset of each chunk.
-        //  2. Build a list of chunks, and (a) use it for rewinding when restarting at last
-        //     committed offset (per call to getCommittedWriteSize); (b) remove chunks from the list
-        //     that have been persisted.
-        //  3. Limit the list size to some number (possibly flag-controlled) of sent chunks.
-        throw new IOException(
-            String.format("Insert failed for '%s'", resourceId), responseObserver.getError());
-      }
+        if (responseObserver.hasError()) {
+          // If error happens, wrap the source input stream with PushbackInputStream so that
+          // latest read data chunk can be re-read and insert during the next insert call.
+          this.pushbackPipeSource = new PushbackInputStream(pipeSource, MAX_BYTES_PER_MESSAGE);
```

Comment:
I think it will be simpler and less error-prone to use [`BufferedInputStream`](https://docs.oracle.com/javase/8/docs/api/java/io/BufferedInputStream.html) with [`mark`](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#mark-int-)/[`reset`](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#reset--) methods instead of manually maintaining retry cache with `PushbackInputStream`.

Reply:
I considered this option, mark/reset seems to be favorable if entire data chunk from last insert needs to be committed again. However, I didn't figure out how to use mark/reset handles the case where only partial of the last data chunk was committed. With pushbackInputStream I can achieve this by unreading length of uncommitted data from pushbackBuffer and there doesn't seem to be a way to do this easily with BufferedInputStream.

Reply:
To re-read part of the marked data, you can use [`skip`](https://docs.oracle.com/javase/8/docs/api/java/io/InputStream.html#skip-long-) method to skip n-bytes to the 1st byte that you want to re-read and continue read from there.

Reply:
Updated. I overlooked the skip() method. It does simplify the implementation.